src/osparc_control/models.py

```diff
@@ -0,0 +1,124 @@
+from enum import Enum
+from typing import Any
+from typing import Dict
+from typing import List
+from typing import Optional
+
+import umsgpack  # type: ignore
+from pydantic import BaseModel
+from pydantic import Extra
+from pydantic import Field
+from pydantic import PrivateAttr
+from pydantic import validator
+
+
+class CommandBase(BaseModel):
+    def to_bytes(self) -> bytes:
+        return umsgpack.packb(self.dict())  # type: ignore
+
+    @classmethod
+    def from_bytes(cls, raw: bytes) -> Optional[Any]:
+        return cls.parse_obj(umsgpack.unpackb(raw))
+
+    class Config:
+        extra = Extra.allow
+
+
+class CommandParameter(BaseModel):
+    name: str = Field(
+        ..., description="name of the parameter to be provided with the commnad"
+    )
+    description: str = Field(
+        ...,
+        description=(
+            "provide more information to the user, how this command should be used"
+        ),
+    )
+
+
+class CommnadType(str, Enum):
+    # the command expects no reply
+    WITHOUT_REPLY = "WITHOUT_REPLY"
+
+    # the command will provide a reply
+    # the user is require to check for the results
+    # of this reply
+    WITH_DELAYED_REPLY = "WITH_DELAYED_REPLY"
+
+    # the command will return the result immediately
+    # and user code will be blocked until reply arrives
+    # used for very fast replies (provide data which already exists)
+    WITH_IMMEDIATE_REPLY = "WITH_IMMEDIATE_REPLY"
+
+
+class CommandManifest(BaseModel):
+    # used internally
+    _remapped_params: Dict[str, CommandParameter] = PrivateAttr()
+
+    action: str = Field(..., description="name of the action to be triggered on remote")
+    description: str = Field(..., description="more details about the action")
+    params: List[CommandParameter] = Field(
+        None, description="requested parameters by the user"
+    )
+    command_type: CommnadType = Field(
+        ..., description="describes the command type, behaviour and usage"
+    )
+
+    @validator("params")
+    @classmethod
+    def ensure_unique_parameter_names(
+        cls, v: List[CommandParameter]
+    ) -> List[CommandParameter]:
+        if len(v) != len({x.name for x in v}):
+            raise ValueError(f"Duplicate CommandParameter name found in {v}")
```

Comment:
you might want to display which parameter it is as well

Reply:
I think it's not worth the effort since it will be displayed twice and will be simple enough to debug.